plugin/rewrite/name.go

```diff
@@ -16,6 +18,36 @@ func newNameRule(from, to string) (Rule, error) {
 
 // Rewrite rewrites the the current request.
 func (rule *nameRule) Rewrite(w dns.ResponseWriter, r *dns.Msg) Result {
+	requestName := r.Question[0].Name
+	// Rewrite based on substring match
+	if strings.HasPrefix(rule.From, "sub:") {
+		substituteStr := strings.Replace(rule.From, "sub:", "", -1)
+		if strings.Contains(requestName, substituteStr) {
+			r.Question[0].Name = strings.Replace(requestName, substituteStr, rule.To, -1)
+			return RewriteDone
+		}
+	}
+	// Rewrite based on regular expressions
+	if strings.HasPrefix(rule.From, "regex:") {
+		pattern := strings.Replace(rule.From, "regex:", "", -1)
+		regexPattern, err := regexp.Compile(pattern)
+		if err != nil {
+			return RewriteIgnored
+		}
+		regexGroups := regexPattern.FindStringSubmatch(requestName)
+		if len(regexGroups) == 0 {
+			return RewriteIgnored
+		}
+		for groupIndex, groupValue := range regexGroups {
+			groupIndexStr := "{" + strconv.Itoa(groupIndex) + "}"
+			if strings.Contains(rule.To, groupIndexStr) {
+				rule.To = strings.Replace(rule.To, groupIndexStr, groupValue, -1)
+			}
+		}
+		rule.From = r.Question[0].Name
```

Comment:
Why do you change the rule.From here? To trigger the "exact match" code? You can't change the rule at this point (which you mention in your latest comment).

You can move the exact match code into a separate function and then call it from here as well as from the matchType `exact` (better constant name than `full` I said above). In fact each matchType probably should be in a function to make this a simple switch statement.

Reply:
> You can move the exact match code into a separate function and then call it from here as well as from the matchType exact (better constant name than full I said above). In fact each matchType probably should be in a function to make this a simple switch statement.

👍